### PR TITLE
NIAD-1196: Map SpecimenCharacteristic identification

### PR DESCRIPTION
**Pull request by andrewwa-kainos**
## Description

NHS002 messages can have SPC segments with 7867 Specimen Characteristic Identification" details and no free text description.

Both are optional.

## Jira Ticket

https://gpitbjss.atlassian.net/browse/NIAD-1196

## Checklist

These are items (excluding GitHub Checks) which should be confirmed before a branch is ready to merge.

- [x] Acceptance Criteria met
- [x] Commit messages are meaningful
- [x] Manually tested
- [x] Self-reviewed your code
- [ ] Code reviewed from two other developers
- [x] If your pull request depends on any other, please link them in the description
- [x] main branch is passing/green on CI
- [x] Add/update any relevant documentation

### CODE REVIEW

File: src/intTest/resources/edifact/pathology_2.edifact.dat [hunk 1]
```diff
@@ -38,7 +38,7 @@ FTX+CID+++TIRED ALL THE TIME, LOW Hb'
 FTX+CID+++COELIAC'
 S16+16'
 SEQ++1'
-SPC+TSP+:::VENOUS BLOOD'
+SPC+TSP+T016:920'
```
Comment (n-r-b) [inline src/intTest/resources/edifact/pathology_2.edifact.dat:41]: Is `pathology_2.edifact.dat` for `NHS002` or just `NHS003` pathology messages? Just checking if this is the best place to add `NHS002`  examples

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        final var segment = SpecimenCharacteristic.fromString("SPC+TSP+A012:::Free text");
+
+        assertAll(
+            () -> assertThat(segment.getCharacteristic()).contains("A012"),
+            () -> assertThat(segment.getTypeOfSpecimen()).contains("Free text")
+        );
+    }
+
+    @Test
+    void testValidateNoneRequired() {
```
Comment (n-r-b) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java:39]: Is this a valid scenario? Should we at least validate either the `type` or `characteristic` exists?

File: src/intTest/resources/edifact/pathology_2.edifact.dat (hunk 1 above)
Comment (bartek-sarul) [inline src/intTest/resources/edifact/pathology_2.edifact.dat:41]: Can we have this example having both text and code?

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package uk.nhs.digital.nhsconnect.lab.results.model.edifact;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatNoException;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+class SpecimenCharacteristicTest {
```
Comment (n-r-b) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java:9]: Can you add a test for an invalid edifact? i.e. doesn't start with `SPC+TSP`

File: src/intTest/resources/edifact/pathology_2.edifact.dat (hunk 1 above)
Comment (andrewwa-kainos) [reply src/intTest/resources/edifact/pathology_2.edifact.dat:41]: Sure. Is that to be more realistic, or to better show off (test) the range of functionality?

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java [hunk 4]
```diff
@@ -56,12 +55,16 @@ private Specimen edifactToFhir(final SpecimenDetails edifact, final Patient pati
...
-        Optional.ofNullable(edifact.getCharacteristicType())
-            .map(SpecimenCharacteristicType::getTypeOfSpecimen)
-            .ifPresent(specimenType -> fhir.getType().addCoding()
-                .setDisplay(specimenType)
-                .setSystem(SNOMED_CODING_SYSTEM));
+        // fhir.type = SG16.SPC.C832
+        Optional.ofNullable(edifact.getCharacteristic())
+            .ifPresent(specimenCharacteristic -> {
+                final var coding = fhir.getType().addCoding()
+                    .setSystem(SNOMED_CODING_SYSTEM);
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java:62]: Do we know if the system is the same for different characteristic types? Any idea what `903` means and if this affects the system in any way?

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/segmentgroup/SpecimenDetailsTest.java [hunk 5]
```diff
@@ -47,10 +47,10 @@ void testGetSpecimenCharacteristicType() {
             "SPC+TSP+:::BLOOD & URINE",
             "ignore me"
         ));
-        assertThat(specimen.getCharacteristicType())
+        assertThat(specimen.getCharacteristic())
```
Comment (n-r-b) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/segmentgroup/SpecimenDetailsTest.java:50]: Can you add a test for `SPC+TSP+T016:920`?

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java [hunk 6]
```diff
@@ -56,12 +55,16 @@ private Specimen edifactToFhir(final SpecimenDetails edifact, final Patient pati
...
+                specimenCharacteristic.getCharacteristic()
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java:63]: Can you reflect this change in the `SpecimenMapperTest`?

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Segment.java [hunk 7]
```diff
@@ -32,4 +33,12 @@
 
     public void validate() throws EdifactValidationException {
     }
+
+    protected static Optional<String> arrayGetSafe(String[] array, int index) {
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/Segment.java:37]: Should we add a test for this?

File: src/intTest/resources/edifact/pathology_2.edifact.dat (hunk 1 above)
Comment (andrewwa-kainos) [reply src/intTest/resources/edifact/pathology_2.edifact.dat]: `pathology_2.edifact.dat` holds an example NHS002 message (incrementally becoming more and more NHS002ey as we play the stories). `pathology_3.edifact.dat` keeps to the NHS003 expectations.

Comment (n-r-b) [reply src/intTest/resources/edifact/pathology_2.edifact.dat]: Ok, i didn't realise that - thought it was just another example of `NHS003`

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java (hunk 2 above)
Comment (andrewwa-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/SpecimenCharacteristicTest.java:39]: Probably... I'll do it.

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java (hunk 4 above)
Comment (andrewwa-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java:62]: No clue!

**Review by n-r-b — APPROVED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java (hunk 4 above)
Comment (andrewwa-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/SpecimenMapper.java:62]: As we couldn't find this information, we've agreed to keep it as it is and if there's a problem then hopefully the test partners will find it!

File: src/intTest/resources/edifact/pathology_2.edifact.dat (hunk 1 above)
Comment (bartek-sarul) [reply src/intTest/resources/edifact/pathology_2.edifact.dat]: To show both are mapped

**Review by bartek-sarul — APPROVED** [review]

**Pull request merged**